### PR TITLE
Add build workflows; fix macos build errors/warnings; update code formatting

### PR DESCRIPTION
**Pull request by davidlion**
# References
This PR addresses macos build issues that impact https://github.com/y-scope/clp/pull/131.

# Description
This PR addresses a few house keeping issues:
## Building
- create github workflows to build on ubuntu-latest and macos-latest
- address minor macos build issues
## Formatting
- updates clang format and tidy configurations
- applies clang format to all files
- reflows comments to be 80 characters

# Validation performed
All CLP unittests passing using https://github.com/y-scope/clp/pull/131.

### CODE REVIEW

File: src/log_surgeon/LALR1Parser.tpp [hunk 1]
```diff
@@ -2,17 +2,58 @@
...
-
-// C++ standard libraries
 #include <functional>
 #include <iostream>
 #include <optional>
 
-// Project headers
-#include "Constants.hpp"
+#include <log_surgeon/Constants.hpp>
+#include <log_surgeon/LALR1Parser.hpp>
```
Comment (SharafMohamed) [inline src/log_surgeon/LALR1Parser.tpp:10]:
```suggestion

```

File: src/log_surgeon/LALR1Parser.tpp [hunk 2]
```diff
@@ -141,28 +192,33 @@ void LALR1Parser<NFAStateType, DFAStateType>::generate_lr0_kernels() {
...
 }
 
 template <typename NFAStateType, typename DFAStateType>
 void LALR1Parser<NFAStateType, DFAStateType>::generate_lr0_closure(ItemSet* item_set_ptr) {
-    std::deque<Item> q(item_set_ptr->m_kernel.begin(),
-                       item_set_ptr->m_kernel.end()); // {{S'->(dot)S, ""}}
+    std::deque<Item> q(
+            item_set_ptr->m_kernel.begin(),
+            item_set_ptr->m_kernel.end()
+    );  // {{S'->(dot)S, ""}}
```
Comment (SharafMohamed) [inline src/log_surgeon/LALR1Parser.tpp:219]:
```suggestion
    // add {{S'->(dot)S, ""}}
    std::deque<Item> q(
            item_set_ptr->m_kernel.begin(),
            item_set_ptr->m_kernel.end()
    ); 
```

File: src/log_surgeon/Lexer.tpp [hunk 3]
```diff
@@ -25,8 +23,8 @@ void Lexer<NFAStateType, DFAStateType>::flip_states(uint32_t old_storage_size) {
     } else {
         m_match_pos += old_storage_size / 2;
     }
-    /// TODO when m_start_pos == old_storage_size / 2, theres two possible cases currently so both
-    /// options are potentially wrong
+    // TODO when m_start_pos == old_storage_size / 2, theres two possible cases
+    // currently so both / options are potentially wrong
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:27]:
```suggestion
    // currently so both options are potentially wrong
```

File: src/log_surgeon/Lexer.tpp [hunk 4]
```diff
@@ -51,12 +49,13 @@ auto Lexer<NFAStateType, DFAStateType>::scan(ParserInputBuffer& input_buffer, To
...
-                          input_buffer.storage().size(),
-                          m_match_line,
-                          m_type_ids};
+            token
+                    = Token{m_start_pos,
+                            m_match_pos,
+                            input_buffer.storage().get_active_buffer(),
+                            input_buffer.storage().size(),
+                            m_match_line,
+                            m_type_ids};
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:58]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 5]
```diff
@@ -97,53 +97,59 @@ auto Lexer<NFAStateType, DFAStateType>::scan(ParserInputBuffer& input_buffer, To
...
-                                  input_buffer.storage().size(),
-                                  m_last_match_line,
-                                  &cTokenUncaughtStringTypes};
+                    token
+                            = Token{m_last_match_pos,
+                                    m_start_pos,
+                                    input_buffer.storage().get_active_buffer(),
+                                    input_buffer.storage().size(),
+                                    m_last_match_line,
+                                    &cTokenUncaughtStringTypes};
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:106]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 6]
```diff
@@ -97,53 +97,59 @@ auto Lexer<NFAStateType, DFAStateType>::scan(ParserInputBuffer& input_buffer, To
...
-                              input_buffer.storage().size(),
-                              m_match_line,
-                              m_type_ids};
+                token
+                        = Token{m_start_pos,
+                                m_match_pos,
+                                input_buffer.storage().get_active_buffer(),
+                                input_buffer.storage().size(),
+                                m_match_line,
+                                m_type_ids};
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:118]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 7]
```diff
@@ -97,53 +97,59 @@ auto Lexer<NFAStateType, DFAStateType>::scan(ParserInputBuffer& input_buffer, To
...
-                                  input_buffer.storage().size(),
-                                  m_last_match_line,
-                                  &cTokenUncaughtStringTypes};
+                    token
+                            = Token{m_last_match_pos,
+                                    m_start_pos,
+                                    input_buffer.storage().get_active_buffer(),
+                                    input_buffer.storage().size(),
+                                    m_last_match_line,
+                                    &cTokenUncaughtStringTypes};
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:132]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 8]
```diff
@@ -97,53 +97,59 @@ auto Lexer<NFAStateType, DFAStateType>::scan(ParserInputBuffer& input_buffer, To
...
-                              input_buffer.storage().size(),
-                              m_line,
-                              &cTokenEndTypes};
+                token
+                        = Token{input_buffer.storage().pos(),
+                                input_buffer.storage().pos(),
+                                input_buffer.storage().get_active_buffer(),
+                                input_buffer.storage().size(),
+                                m_line,
+                                &cTokenEndTypes};
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:141]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 9]
```diff
@@ -172,12 +180,13 @@ auto Lexer<NFAStateType, DFAStateType>::scan_with_wildcard(ParserInputBuffer& in
...
-                          input_buffer.storage().size(),
-                          m_match_line,
-                          m_type_ids};
+            token
+                    = Token{m_start_pos,
+                            m_match_pos,
+                            input_buffer.storage().get_active_buffer(),
+                            input_buffer.storage().size(),
+                            m_match_line,
+                            m_type_ids};
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:189]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 10]
```diff
@@ -215,12 +225,13 @@ auto Lexer<NFAStateType, DFAStateType>::scan_with_wildcard(ParserInputBuffer& in
...
-                              input_buffer.storage().size(),
-                              m_last_match_line,
-                              &cTokenUncaughtStringTypes};
+                token
+                        = Token{m_last_match_pos,
+                                input_buffer.storage().pos(),
+                                input_buffer.storage().get_active_buffer(),
+                                input_buffer.storage().size(),
+                                m_last_match_line,
+                                &cTokenUncaughtStringTypes};
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:234]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 11]
```diff
@@ -229,12 +240,13 @@ auto Lexer<NFAStateType, DFAStateType>::scan_with_wildcard(ParserInputBuffer& in
...
-                                          input_buffer.storage().size(),
-                                          m_last_match_line,
-                                          &cTokenUncaughtStringTypes};
+                            token
+                                    = Token{m_last_match_pos,
+                                            input_buffer.storage().pos(),
+                                            input_buffer.storage().get_active_buffer(),
+                                            input_buffer.storage().size(),
+                                            m_last_match_line,
+                                            &cTokenUncaughtStringTypes};
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:249]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 12]
```diff
@@ -245,12 +257,13 @@ auto Lexer<NFAStateType, DFAStateType>::scan_with_wildcard(ParserInputBuffer& in
...
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:266]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 13]
```diff
@@ -271,12 +284,13 @@ auto Lexer<NFAStateType, DFAStateType>::scan_with_wildcard(ParserInputBuffer& in
...
-                              m_type_ids};
+                token
+                        = Token{m_start_pos,
+                                m_match_pos,
+                                input_buffer.storage().get_active_buffer(),
+                                input_buffer.storage().size(),
+                                m_match_line,
+                                m_type_ids};
                 return ErrorCode::Success;
             }
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:295]: Formatting is weird

File: src/log_surgeon/Lexer.tpp [hunk 14]
```diff
@@ -271,12 +284,13 @@ auto Lexer<NFAStateType, DFAStateType>::scan_with_wildcard(ParserInputBuffer& in
...
-                              input_buffer.storage().size(),
-                              m_match_line,
-                              m_type_ids};
+                token
+                        = Token{m_start_pos,
+                                m_match_pos,
+                                input_buffer.storage().get_active_buffer(),
+                                input_buffer.storage().size(),
+                                m_match_line,
+                                m_type_ids};
```
Comment (SharafMohamed) [inline src/log_surgeon/Lexer.tpp:293]: Formatting is weird

File: src/log_surgeon/ParserInputBuffer.cpp [hunk 15]
```diff
@@ -1,10 +1,10 @@
-// C++ libraries
+#include "ParserInputBuffer.hpp"
+
 #include <memory.h>
+
```
Comment (SharafMohamed) [inline src/log_surgeon/ParserInputBuffer.cpp:4]:
```suggestion
```
unneeded newline?

File: src/log_surgeon/SchemaParser.hpp [hunk 16]
```diff
@@ -40,13 +39,15 @@ class IdentifierAST : public ParserAST {
...
-          m_regex_ptr(std::move(regex_ptr)) {}
+    SchemaVarAST(
+            std::string name,
+            std::unique_ptr<finite_automata::RegexAST<finite_automata::RegexNFAByteState>>
+                    regex_ptr,
+            uint32_t line_num
+    )
+            : m_line_num(line_num),
+              m_name(std::move(name)),
+              m_regex_ptr(std::move(regex_ptr)) {}
```
Comment (SharafMohamed) [inline src/log_surgeon/SchemaParser.hpp:50]: All the initializer lists seem to be starting on the next line

File: src/log_surgeon/SchemaParser.cpp [hunk 17]
```diff
@@ -1,20 +1,19 @@
 #include "SchemaParser.hpp"
 
-// C++ libraries
 #include <cmath>
+
```
Comment (SharafMohamed) [inline src/log_surgeon/SchemaParser.cpp:4]: Unintentional newline here?

File: src/log_surgeon/ParserInputBuffer.cpp [hunk 18]
```diff
@@ -1,10 +1,10 @@
-// C++ libraries
+#include "ParserInputBuffer.hpp"
+
 #include <memory.h>
```
Comment (SharafMohamed) [inline src/log_surgeon/ParserInputBuffer.cpp:3]: Why <memory.h> instead of <memory>

File: src/log_surgeon/ParserInputBuffer.cpp (hunk 15 above)
Comment (SharafMohamed) [inline src/log_surgeon/ParserInputBuffer.cpp:4]: Unintentional newline here?

File: src/log_surgeon/SchemaParser.cpp [hunk 19]
```diff
@@ -23,10 +22,10 @@ using RegexASTLiteralByte = log_surgeon::finite_automata::RegexASTLiteral<
...
 using RegexASTMultiplicationByte = log_surgeon::finite_automata::RegexASTMultiplication<
         log_surgeon::finite_automata::RegexNFAByteState>;
-using RegexASTOrByte =
-        log_surgeon::finite_automata::RegexASTOr<log_surgeon::finite_automata::RegexNFAByteState>;
-using RegexASTCatByte =
-        log_surgeon::finite_automata::RegexASTCat<log_surgeon::finite_automata::RegexNFAByteState>;
+using RegexASTOrByte
+        = log_surgeon::finite_automata::RegexASTOr<log_surgeon::finite_automata::RegexNFAByteState>;
+using RegexASTCatByte = log_surgeon::finite_automata::RegexASTCat<
+        log_surgeon::finite_automata::RegexNFAByteState>;
```
Comment (SharafMohamed) [inline src/log_surgeon/SchemaParser.cpp:28]: Is this exactly 1 character over the length when formatted as the lines above it?

File: src/log_surgeon/SchemaParser.cpp [hunk 20]
```diff
@@ -294,34 +314,41 @@ static auto regex_wildcard_rule(NonTerminal* /* m */) -> unique_ptr<ParserAST> {
...
+            new ParserValueRegex(unique_ptr<RegexASTByte>(new RegexASTLiteralByte('\n')))
+    );
 }
 
 static auto regex_white_space_rule(NonTerminal* /* m */) -> unique_ptr<ParserAST> {
-    unique_ptr<RegexASTGroupByte> regex_ast_group =
-            make_unique<RegexASTGroupByte>(RegexASTGroupByte({' ', '\t', '\r', '\n', '\v', '\f'}));
+    unique_ptr<RegexASTGroupByte> regex_ast_group
+            = make_unique<RegexASTGroupByte>(RegexASTGroupByte({' ', '\t', '\r', '\n', '\v', '\f'})
+            );
```
Comment (SharafMohamed) [inline src/log_surgeon/SchemaParser.cpp:348]: Is this formatted like this cause it does not fit on one line?

Comment (davidlion) [reply src/log_surgeon/SchemaParser.cpp:348]: yup

File: src/log_surgeon/SchemaParser.cpp (hunk 19 above)
Comment (davidlion) [reply src/log_surgeon/SchemaParser.cpp:28]: it is 115, so 15 over

File: src/log_surgeon/SchemaParser.cpp (hunk 17 above)
Comment (davidlion) [reply src/log_surgeon/SchemaParser.cpp:4]: This is from a bug in the .clang-format regex, good catch.

File: src/log_surgeon/ParserInputBuffer.cpp (hunk 15 above)
Comment (SharafMohamed) [inline src/log_surgeon/ParserInputBuffer.cpp:4]:
```suggestion
#include <memory>
```

File: src/log_surgeon/SchemaParser.hpp (hunk 16 above)
Comment (davidlion) [reply src/log_surgeon/SchemaParser.hpp:50]: We're going to live with this case, as it makes functions/method nicer.

**Review by SharafMohamed — APPROVED** [review]

**Review by kirkrodrigues — APPROVED** [review]

**Pull request merged**